F/short subaccount id change

Tofel opened the pull request:
Convert subaccount_id to decimal when **serializing** short subaccount id.

Tofel committed:
chore: convert hexadecimal subaccount id to decimal on creation

Tofel committed:
chore: lints

Tofel committed:
chore: convert short subaccount_id to decimal only when serialising

Tofel committed:
chore: fix lints and tests

gorgos commented on packages/injective-cosmwasm/src/exchange/types.rs:
```diff
@@ -258,6 +272,22 @@ impl<'de> Deserialize<'de> for ShortSubaccountId {
+    where
+        S: Serializer,
+    {
+        // let's check first check again if it's a hexadecimal number and parse to decimal (expected by the chain)
```
```suggestion
        // chain expects decimal number, but ShortSubaccountId stores hexadecimal string
```

gorgos commented on packages/injective-cosmwasm/src/exchange/types.rs:
```diff
@@ -258,6 +272,22 @@ impl<'de> Deserialize<'de> for ShortSubaccountId {
+        // let's check first check again if it's a hexadecimal number and parse to decimal (expected by the chain)
+        let as_decimal = match u32::from_str_radix(self.0.as_str(), 16) {
+            Ok(dec) => Ok(dec),
+            Err(_) => Err(S::Error::custom("Invalid value: ShortSubaccountId was not a hexadecimal number")),
```
why not add the `self.0` to the error?

gorgos commented on packages/injective-cosmwasm/src/exchange/types.rs:
```diff
@@ -258,6 +272,22 @@ impl<'de> Deserialize<'de> for ShortSubaccountId {
+        let as_decimal = match u32::from_str_radix(self.0.as_str(), 16) {
+            Ok(dec) => Ok(dec),
+            Err(_) => Err(S::Error::custom("Invalid value: ShortSubaccountId was not a hexadecimal number")),
+        };
```
I would add the `?` here.

gorgos commented on packages/injective-cosmwasm/src/exchange/types.rs:
```diff
@@ -258,6 +272,22 @@ impl<'de> Deserialize<'de> for ShortSubaccountId {
+            Err(_) => Err(S::Error::custom("Invalid value: ShortSubaccountId was not a hexadecimal number")),
+        };
+
+        // if the string cannot be parsed to a decimal return error otherwise let's pad it with left zeros
```
unnecessary comment

gorgos commented on packages/injective-cosmwasm/src/exchange/types.rs:
```diff
@@ -218,7 +219,20 @@ impl ShortSubaccountId {
         let id = id_s.into();
 
-        match id.parse::<u16>() {
+        // let's check first if it's a hexadecimal number and parse to decimal (expected by the chain)
+        let as_decimal = match u32::from_str_radix(id.as_str(), 16) {
+            Ok(dec) => Ok(dec),
+            Err(_) => Err(StdError::generic_err("Invalid value: ShortSubaccountId was not a hexadecimal number")),
+        };
+
+        // if the string cannot be parsed to a decimal return error
+        let decimal_u32 = as_decimal?;
+
+        // if it is a hexadecimal number, let's pad it with left zeros
+        let number_to_use = format!("{:0>3}", decimal_u32.to_string());
+
+        // check if resulting number isn't bigger than chain limit
+        match number_to_use.parse::<u16>() {
```
I don't think we need this, do we? if we already do the conversion to decimal during serializations.

- If we do, please say why and remove the comments.
- If we don't, please remove here.

Tofel replied:
Well, I think we should validate the input, when a new instance is created. You don't want to find out that input is invalid only when you're serializing it, which would probably when submessage is dispatched? 

This part could deffo be removed here:
```
       // if it is a hexadecimal number, let's pad it with left zeros
        let number_to_use = format!("{:0>3}", decimal_u32.to_string());
```

Also, I will make that code more concise. Plus, add a method `must_new()` which will panic on invalid input. I that once this PR is merged we should change all usages of `unchecked()` for `must_new()`: https://github.com/InjectiveLabs/cw-injective/issues/160

Tofel committed:
Update packages/injective-cosmwasm/src/exchange/types.rs

Co-authored-by: Markus Waas <mail@markuswaas.de>

Tofel committed:
chore: CR comments

Tofel committed:
Merge branch 'f/short_subaccount_id_change' of github.com:InjectiveLabs/cw-injective into f/short_subaccount_id_change

Tofel committed:
Merge branch 'dev' into f/short_subaccount_id_change

Tofel committed:
chore: add suport for deserializing short subaccount id from long

Tofel committed:
chore: remove commented out code

Tofel committed:
chore: use validate also when serializing, adjust tests

Tofel commented:
@gorgos added support for deserializing from long subaccount id, added some tests and added validate function (that's now used by `new()`, `must_new()` and `serialize()` and `deserialize()`.

gorgos reviewed (changes requested):
The changes LGTM, but I thought we discussed also adding a new feature to deserialize a long subaccount id into the short type?
